clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,287 @@
+            .setRebalanceTimeoutMs(rebalanceTimeoutMs);
+
+        membershipManager.groupInstanceId().ifPresent(data::setInstanceId);
+
+        if (this.subscriptions.hasPatternSubscription()) {
+            // We haven't discsussed how Regex is stored in the consumer. We could do it in the subscriptionState
+            // , in the memberStateManager, or here.
+            // data.setSubscribedTopicRegex(regex)
+        } else {
+            data.setSubscribedTopicNames(new ArrayList<>(this.subscriptions.subscription()));
+        }
+
+        this.membershipManager.assignorSelection().serverAssignor().ifPresent(data::setServerAssignor);
+
+        NetworkClientDelegate.UnsentRequest request = new NetworkClientDelegate.UnsentRequest(
+            new ConsumerGroupHeartbeatRequest.Builder(data),
+            coordinatorRequestManager.coordinator());
+
+        request.future().whenComplete((response, exception) -> {
+            if (exception == null) {
+                onResponse((ConsumerGroupHeartbeatResponse) response.responseBody(), response.receivedTimeMs());
+            } else {
+                onFailure(exception, response.receivedTimeMs());
+            }
+        });
+        return request;
+    }
+
+    private void onFailure(final Throwable exception, final long responseTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(responseTimeMs);
+        logger.debug("failed sending heartbeat due to {}", exception.getMessage());
+    }
+
+    private void onResponse(final ConsumerGroupHeartbeatResponse response, long currentTimeMs) {
+        if (response.data().errorCode() == Errors.NONE.code()) {
+            this.heartbeatRequestState.updateHeartbeatIntervalMs(response.data().heartbeatIntervalMs());
+            this.heartbeatRequestState.onSuccessfulAttempt(currentTimeMs);
+            this.heartbeatRequestState.reset();
+            try {
+                membershipManager.updateState(response.data());
+            } catch (KafkaException e) {
+                logger.error("Received unexpected error in heartbeat response: {}", e.getMessage());
+            }
+            return;
+        }
+
+        onErrorResponse(response, currentTimeMs);
+    }
+
+    private void onErrorResponse(final ConsumerGroupHeartbeatResponse response,
+                                 final long currentTimeMs) {
+        this.heartbeatRequestState.onFailedAttempt(currentTimeMs);
+        short errorCode = response.data().errorCode();
+        if (errorCode == Errors.NOT_COORDINATOR.code() || errorCode == Errors.COORDINATOR_NOT_AVAILABLE.code()) {
+            logInfo("Coordinator is either not started or not valid. Retrying", response, currentTimeMs);
+            coordinatorRequestManager.markCoordinatorUnknown(response.data().errorMessage(), currentTimeMs);
+        } else if (errorCode == Errors.COORDINATOR_LOAD_IN_PROGRESS.code()) {
+            // retry
+            logInfo("Coordinator {} is loading. Retrying", response, currentTimeMs);
```

Comment:
```suggestion
            logInfo(String.format("Heartbeat was not successful because the coordinator node %s is loading. Retrying", coordinatorRequestManager.coordinator()), response, currentTimeMs);
```